### DISCUSSION TITLE
Clarifications on the erasure code calculator

### DISCUSSION THREAD

**Discussion by Recsy** [minio/minio#13902] [Q&A] [answered]
Hello!

I have been playing around with the MinIO erasure code calculator, and I have some questions to hopefully help me understand how it works.

The default configuration on the calculator is 8 servers and 16 drives (8x16). That defaults to a stripe size of 16, and a storage efficiency of 75%. If I change the number of servers to 9, the stripe size is locked at 9 and the storage efficiency drops to 55%. Can you help me understand why only the stripe size of 9 is an option here? I'm wondering because it means that the usable capacity of the 8 server setup is actually higher, so it would appear that 8x16 is a superior setup.

Another question I have is how I should view this efficiency when I decide to add pools to my cluster. If I start with a 4x16 setup (75% efficiency), and add a 6x16 pool (66% efficiency) to the cluster, what is the storage efficiency of objects inserted after this point?

**Comment by klauspost** [comment, marked as answer]
>  If I change the number of servers to 9

It will distribute sets on servers equally. With 8 servers it will be 2 disks/set on each server. Changing to 9 servers will use one disk from each per set. You can lower the "Erasure Code Parity (M)" if you want to have less erasure codes.

>  If I start with a 4x16 setup (75% efficiency), and add a 6x16 pool (66% efficiency) to the cluster, what is the storage efficiency of objects inserted after this point?

This depends on which server pool the object lands on. Objects are distributed between pools with the ratio of free disk space, so it is to some extent random, but it will be one or the other.

**Reply by Recsy** [reply]
Thanks @klauspost, that definitely clarified some things.

I am curious on one more thing, that may not warrant a discussion of its own. How does MinIO distribute stuff around and ensure that the disks are evenly balanced? If I start with N empty disks, will they fill up around the same time, or can I run into some issues with disks filling up before others?

**Reply by klauspost** [reply]
Files on disks in a set are all the same size.

With multiple sets objects are distributed on hash. For many objects this will even out, but sets with few, huge files, sets may fill before others.

As mentioned, between pools we actively decide where objects are placed.
